src/main.rs

```diff
@@ -3,10 +3,105 @@ use clap::{Arg, Command};
 use linkme::distributed_slice;
 use metriken::Lazy;
 use ringlog::*;
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::time::SystemTime;
+use tokio::sync::RwLock;
 
 type Duration = clocksource::Duration<clocksource::Nanoseconds<u64>>;
 type Instant = clocksource::Instant<clocksource::Nanoseconds<u64>>;
 
+type HistogramSnapshots = HashMap<String, metriken::histogram::Snapshot>;
+
+static SNAPSHOTS: Lazy<Arc<RwLock<Snapshots>>> =
+    Lazy::new(|| Arc::new(RwLock::new(Snapshots::new())));
+
+pub struct Snapshots {
+    timestamp: SystemTime,
+    previous: HistogramSnapshots,
+    deltas: HistogramSnapshots,
+}
+
+impl Default for Snapshots {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Snapshots {
+    pub fn new() -> Self {
+        let timestamp = SystemTime::now();
+
+        let mut current = HashMap::new();
+
+        for metric in metriken::metrics().iter() {
+            let any = if let Some(any) = metric.as_any() {
+                any
+            } else {
+                continue;
+            };
+
+            let key = metric.name().to_string();
+
+            if let Some(histogram) = any.downcast_ref::<metriken::AtomicHistogram>() {
```

Comment:
Not sure if the type system will allow it, but since both Atomic and RwLockHistogram return the same kind of snapshot, can the two arms be combined with an ||?

Reply:
Not unless we refactor to do something with traits. I can refactor this callsite slightly, but it's not a huge improvement.

Reply:
Made a small change here.

Reply:
Don't have much of an opinion on which pattern is better, but we may as well use the same one in `new` and `update`.